packages/internal/src/utils.js

```diff
@@ -52,6 +52,48 @@ export const deeplyFulfilledObject = async obj => {
   return deeplyFulfilled(obj);
 };
 
+/**
+ * @param {any} value
+ * @param {string | undefined} name
+ * @param {object | undefined} container
+ * @param {(value: any, name: string, record: object) => any} mapper
+ * @returns {any}
+ */
+const deepMapObjectInternal = (value, name, container, mapper) => {
+  if (container && typeof name === 'string') {
+    const mapped = mapper(value, name, container);
+    if (mapped !== value) {
+      return mapped;
+    }
+  }
+
+  if (typeof value !== 'object' || !value) {
+    return value;
+  }
+
+  let wasMapped = false;
+  const mappedEntries = Object.entries(value).map(([innerName, innerValue]) => {
+    const mappedInnerValue = deepMapObjectInternal(
+      innerValue,
+      innerName,
+      value,
+      mapper,
+    );
+    wasMapped ||= mappedInnerValue !== innerValue;
+    return [innerName, mappedInnerValue];
+  });
+
+  return wasMapped ? Object.fromEntries(mappedEntries) : value;
+};
+
+/**
+ * @param {object} obj
```

Comment:
please include some documentation.

a parameterized typedef would be nice but not worth the effort yet imo

Reply:
I couldn't think how a parametrized typedef would work in this case. At best we could pretend that the output type is the same as the input type.

Reply:
I was thinking it would maintain the structure but the return type of each leaf would be widened to the return type of the mapping function. (Since the type system doesn't know more about what the function will do)

But still not worth it